lib/rules/jsx-max-depth.js

```diff
@@ -0,0 +1,151 @@
+/**
+ * @fileoverview Validate JSX maximum depth
+ * @author Chris<wfsr@foxmail.com>
+ */
+'use strict';
+
+let variableUtil = require('../util/variable');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Validate JSX maximum depth',
+      category: 'Stylistic Issues',
+      recommended: false
+    },
+    schema: [{
+      oneOf: [
+        {
+          type: 'integer',
+          minimum: 0
+        },
+        {
+          type: 'object',
+          properties: {
+            max: {
+              type: 'integer',
+              minimum: 0
+            }
+          }
+        }
+
+      ]
+    }]
+  },
+  create: function(context) {
+    const MESSAGE = 'Expected the depth of JSX Elements nested should be {{needed}} but found {{gotten}}.';
+    const DEFAULT_DEPTH = 3;
+
+    let option = context.options[0] || DEFAULT_DEPTH;
+    let maxDepth = (typeof option === 'number' ? option : option.max) || DEFAULT_DEPTH;
+
+    function isJSXElement(node) {
+      return node.type === 'JSXElement';
+    }
+
+    function isExpression(node) {
+      return node.type === 'JSXExpressionContainer';
+    }
+
+    function hasJSX(node) {
+      return isJSXElement(node) || isExpression(node) && isJSXElement(node.expression);
+    }
+
+    function isLeaf(node) {
+      let children = node.children;
+
+      return !children.length || !children.some(hasJSX);
+    }
+
+    function getDepth(node) {
+      let count = 1;
+
+      while (isJSXElement(node.parent) || isExpression(node.parent)) {
+        node = node.parent;
+        if (isJSXElement(node)) {
+          count++;
+        }
+      }
+
+      return count;
+    }
+
+
+    function report(node, depth) {
+      context.report({
+        node: node,
+        message: MESSAGE,
+        data: {
+          gotten: depth,
+          needed: maxDepth
+        }
+      });
+    }
+
+    function findJSXElement(variables, name) {
+      let find = function(refs) {
+        let i = refs.length;
+
+        while (--i >= 0) {
+          if ('writeExpr' in refs[i]) {
```

Comment:
please use the `has` package rather than `in`, to avoid traversing the prototype chain.